src/engine/qcommon/common.cpp

```diff
@@ -861,26 +861,49 @@ void Com_Frame()
 		minMsec = 1;
 	}
 
+	int beforeProcessing = Sys::Milliseconds();
+
 	Com_EventLoop();
-	com_frameTime = Sys::Milliseconds();
 
-	if ( lastTime > com_frameTime )
-	{
-		lastTime = com_frameTime; // possible on first frame
-	}
+	// It must be called at least once.
+	IN_Frame();
 
-	msec = com_frameTime - lastTime;
+	int afterProcessing = Sys::Milliseconds();
+
+	com_frameTime = afterProcessing;
 
-	IN_Frame(); // must be called at least once
+	// lastTime can be greater than com_frameTime on first frame.
+	lastTime = std::min( lastTime, com_frameTime );
+
+	msec = com_frameTime - lastTime;
 
 	while ( msec < minMsec )
 	{
-		//give cycles back to the OS
-		Sys::SleepFor(std::chrono::milliseconds(std::min(minMsec - msec, 50)));
-		IN_Frame();
+		int processingTime = afterProcessing - beforeProcessing;
+
+		/* Fill as much as input frames as possible within this timeframe to not
+		let the game process events in a way it would last longer than that. */
+		int jamTime = std::max( afterProcessing, 50 );
+
+		int sleep = minMsec - msec;
```

Comment:
This is called `sleep`, but we don't actually sleep for this time. Maybe we should call it something else